### PR TITLE
Migrate Workspace reset & edit to use diagnostics

### PR DESCRIPTION
**Pull request by hartbit**
Added an override of `DiagnosticsEngine.wrap` that helps wrapping Void returning throwing closures.

### CODE REVIEW

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Comment by hartbit** [comment]
@swift-ci please smoke test

File: Sources/Workspace/Workspace.swift [hunk 1]
```diff
@@ -335,8 +335,15 @@ public class Workspace {
...
 
     /// Resets the entire workspace by removing the data directory.
-    public func reset() throws {
-        try managedDependencies.reset()
+    ///
+    /// - Parameters:
+    ///     - diagnostics: The diagnostics engine that reports errors, warnings
+    ///       and notes.
+    public func reset(with diagnostics: DiagnosticsEngine) {
+        guard diagnostics.wrap({ try managedDependencies.reset() }) else {
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: It is weird that this reset method throws and others don't.

File: Sources/Workspace/Workspace.swift [hunk 2]
```diff
@@ -373,20 +385,29 @@ public class Workspace {
...
-                baseURL: dependency.repository.url,
-                manifestVersion: toolsVersion.manifestVersion)
+            guard let manifest = diagnostics.wrap({
+                try manifestLoader.load(
+                    package: destination,
+                    baseURL: dependency.repository.url,
+                    manifestVersion: toolsVersion.manifestVersion)
+            }) else {
+                return
+            }
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: We now have a helper method to load the manifest in workspace, we should let it accept a diagnostics manager and use that here instead.

File: Sources/Workspace/Workspace.swift [hunk 3]
```diff
@@ -398,44 +419,77 @@ public class Workspace {
...
+                try repositoryManager.lookupSynchronously(repository: dependency.repository)
+            }) else {
+                return
+            }
+            
+            guard let repo = diagnostics.wrap({
+                try handle.open()
+            }) else {
+                return
+            }
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: We can combine these, right?

```swift
let repo = diagnostics.wrap({ 
    try repositoryManager.lookupSynchronously(repository: dependency.repository)
    return try handle.open()
})
guard let repo = repo else { return }
```

File: Sources/Workspace/Workspace.swift [hunk 4]
```diff
@@ -398,44 +419,77 @@ public class Workspace {
...
         }
 
         // Save the new state.
         managedDependencies[dependency.repository] = dependency.editedDependency(
             subpath: RelativePath(packageName), unmanagedPath: path)
-        try managedDependencies.saveState()
+        
+        guard diagnostics.wrap({ try managedDependencies.saveState() }) else {
+            return
+        }
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: I think we're better off creating two methods for edit, one throwing and one non throwing... that will remove the annoying guards everywhere

File: Sources/Workspace/Workspace.swift (hunk 1 above)
Comment (hartbit) [inline Sources/Workspace/Workspace.swift]: What "others" are you talking about?

Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: The methods that are called below this. (Comment is not related to this PR).

Comment (hartbit) [inline Sources/Workspace/Workspace.swift]: Yeah, thats because `managedDependencies.reset()` ends up calling `FileSystem.writeFileContents` which throws, but `repositoryManager.reset()` calls `FileSystem.removeFileTree` which doesn't throw. Don't know what you want to do about that.

Sounds weird because `FileManager.removeItem(atPath:)` could potentially throw (I think) if a file under that path is opened by another process. No? perhaps `FileSystem.removeFileTree` should propagate that error.

File: Tests/WorkspaceTests/WorkspaceTests.swift [hunk 5]
```diff
@@ -560,10 +566,15 @@ final class WorkspaceTests: XCTestCase {
...
-            } catch _ as WorkspaceDiagnostics.DependencyAlreadyInEditMode {}
+            workspace.edit(
+                dependency: editedDependency,
+                packageName: aManifest.name,
+                diagnostics: diagnostics,
+                revision: dependency.checkoutState!.revision)
+            XCTAssert(diagnostics.hasErrors)
+            XCTAssert(diagnostics.diagnostics.contains(where: {
+                $0.id == WorkspaceDiagnostics.DependencyAlreadyInEditMode.id
+            }))
```
Comment (aciidb0mb3r) [inline Tests/WorkspaceTests/WorkspaceTests.swift]: Off topic: We should get some assert helpers for diagnostics engine.
`XCTAssertNoErrors(diagnostics)`
`XCTAssertContains(diagnostics, WorkspaceDiagnostics.DependencyAlreadyInEditMode.id)`

File: Sources/Workspace/Workspace.swift (hunk 1 above)
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: Yeah, I think we should either make these APIs throwing or return a bool perhaps.
PS: This is not related to this PR, just a general comment.

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test
